src/passes/DeadArgumentElimination.cpp

```diff
@@ -176,50 +189,82 @@ struct DAE : public Pass {
   bool optimize = false;
 
   void run(Module* module) override {
+    DAEFunctionInfoMap infoMap;
+    // Ensure all entries exist so the parallel threads don't modify the data
+    // structure.
+    for (auto& func : module->functions) {
+      infoMap[func->name];
+    }
+    // The null name represents module-level code (not in a function).
+    infoMap[Name()];
+
     // Iterate to convergence.
     while (1) {
-      if (!iteration(module)) {
+      if (!iteration(module, infoMap)) {
         break;
       }
     }
   }
 
-  bool iteration(Module* module) {
+  bool iteration(Module* module, DAEFunctionInfoMap& infoMap) {
     allDroppedCalls.clear();
 
-    DAEFunctionInfoMap infoMap;
-    // Ensure they all exist so the parallel threads don't modify the data
-    // structure.
-    for (auto& func : module->functions) {
-      infoMap[func->name];
+#if DAE_DEBUG
+    // Enable this path to mark all contents as stale at the start of each
+    // iteration, which can be used to check for staleness bugs. Note, though,
+    // that staleness bugs can easily cause serious issues with validation (e.g.
+    // if data is stale we may miss that there is an additional caller, that
+    // prevents refining argument types etc.), so this may not be terribly
+    // helpful.
+    if (getenv("ALWAYS_MARK_STALE")) {
+      for (auto& [_, info] : infoMap) {
+        info.markStale();
+      }
     }
+#endif
+
     DAEScanner scanner(&infoMap);
     scanner.walkModuleCode(module);
-    for (auto& curr : module->exports) {
-      if (curr->kind == ExternalKind::Function) {
-        infoMap[curr->value].hasUnseenCalls = true;
-      }
-    }
     // Scan all the functions.
     scanner.run(getPassRunner(), module);
     // Combine all the info.
+    struct CallContext {
+      Call* call;
+      Function* func;
+    };
     std::map<Name, std::vector<Call*>> allCalls;
     std::unordered_set<Name> tailCallees;
-    for (auto& [_, info] : infoMap) {
+    std::unordered_set<Name> hasUnseenCalls;
+    // Track the function in which relevant expressions exist. When we modify
+    // those expressions we will need to mark the function's info as stale.
+    std::unordered_map<Expression*, Name> expressionFuncs;
+    for (auto& [func, info] : infoMap) {
       for (auto& [name, calls] : info.calls) {
         auto& allCallsToName = allCalls[name];
         allCallsToName.insert(allCallsToName.end(), calls.begin(), calls.end());
+        for (auto* call : calls) {
+          expressionFuncs[call] = func;
+        }
       }
       for (auto& callee : info.tailCallees) {
         tailCallees.insert(callee);
       }
-      for (auto& [name, calls] : info.droppedCalls) {
-        allDroppedCalls[name] = calls;
+      for (auto& [call, dropp] : info.droppedCalls) {
+        allDroppedCalls[call] = dropp;
+      }
+      for (auto& name : info.hasUnseenCalls) {
+        hasUnseenCalls.insert(name);
+      }
```

Comment:
Why are we copying the set in the info into a local set?

Reply:
We computed in parallel information in each function, but it is about other functions. Say when we visited `foo` in parallel then we might have seen an unseen call to `bar` and `baz`. This collecting of it all together gives us the final set of all things with unseen calls.